### PR TITLE
OCSP / Certificate Revocation discussion and testing

### PR DESCRIPTION
**Pull request by nycnewman**
- Certificate Revocation Technote covering certificate revocation methods
- Example Java application to allow protocol traces using Wireshark of OCSP and OCSP Stapling
- Updates to Daml Client App to switch to Java (used with custom Daml build using JDK)
- Run OCSP Responder for Root CA separate to OCSP Responder for Intermediate

### CODE REVIEW

**Review by kareem-DA — APPROVED** [review]
lgtm

File: Documentation/technote-postgresql.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
+[![DAML logo](https://daml.com/wp-content/uploads/2020/03/logo.png)](https://www.daml.com)
```
Comment (cocreature) [inline Documentation/technote-postgresql.md]: includes unrelated commit.

File: sockets/client/SSLSocketClient.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+ * intended for use in the design, construction, operation or
+ * maintenance of any nuclear facility.
+ */
+
+import java.net.*;
+import java.io.*;
+import javax.net.ssl.*;
+
+/*
+ * This example demostrates how to use a SSLSocket as client to
```
Comment (cocreature) [inline sockets/client/SSLSocketClient.java:45]: I’m still not entirely convinced that having a generic Java client & server is sensible in something that claims to show you how to build a secure daml deployment. If you really want to keep it maybe use something more high-level like `HttpsURLConnection`. Manually doing tls handshakes at the socket level seems questionable when it comes to building secure software.

File: run-sandbox.sh [hunk 3]
```diff
@@ -41,10 +41,26 @@ if [ ! -f daml-on-sql-1.10.0.jar ]; then
...
+OCSP_STAPLE="TRUE"
+if [ "$OCSP_STAPLE" == "TRUE" ] ; then
+   export JDK_JAVA_OPTIONS="-javaagent:$ROOTDIR/jSSLKeyLog.jar=server.log -Djava.security.properties=$ROOTDIR/java.security -Dcom.sun.net.ssl.checkRevocation=true -Djdk.tls.client.enableStatusRequestExtension=true -Djdk.tls.server.enableStatusRequestExtension=true -Djava.security.debug='certpath ocsp' -Djavax.net.debug='ssl:handshake,verbose,respmgr'"
+   wget https://github.com/jsslkeylog/jsslkeylog/releases/download/v1.3.0/jSSLKeyLog-1.3.zip -O jSSLKeyLog-1.3.zip
+   unzip -o jSSLKeyLog-1.3.zip jSSLKeyLog.jar
+   JAR="daml-on-sql-binary_deploy.jar"
+   if [ ! -f $JAR ] ; then
+      echo "WARNING: OCSP depends on a custom build of daml-on-sql to use JDK SSLProvider"
+      exit 1
+   fi
```
Comment (cocreature) [inline run-sandbox.sh:53]: Let’s please remove this. Nobody reading this can (due to the lack of instructions) or should (due to the lack of testing & support) be using tghis.

File: Documentation/technote-ocsp-cert-revocation.md [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+## Types of Certificate Revocation Checks
+
+Certificate Revocation mechanisms include:
+
+- **Certificate Revocation Lists (CRL)**
+  - A signed list of Certificates that have been revoked by the CA
+- **OCSP**
+  - Protocol to allow clients of the Certificate Authority to request status of specific certs
+- **OCSP Stapling**
+  - Enhancement to OCSP to allow the server to "staple" the OCSP response removing need for individual clients to check
```
Comment (cocreature) [inline Documentation/technote-ocsp-cert-revocation.md:28]:
```suggestion
  - Enhancement to OCSP to allow the server to "staple" the OCSP response removing need for individual clients to check this in a separate request
```
at least afaik client still check this. the difference is how they get the information they check against.

File: Documentation/technote-ocsp-cert-revocation.md [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+server and waiting on the response during connection setup) and privacy concerns (the CA sees
+all clients who are attempting to access any service they provide certificates for).
+
+**OCSP Stapling** attempts to solve the performance and privacy issues of OCSP. The server
+makes the OCSP call of behalf of all clients connecting to the service and "staples" the response
+in the Server response of the TLS negotiation. In this way, enabled clients receive
+a copy of the response with the certificate and can validate without needing to make OCSP calls themselves.
+This is still prone to availability concerns although the server can cache the OCSP response for a while.
+
+The next two mechanisms wre mainly introduced for browsers and are not directly supported by
```
Comment (cocreature) [inline Documentation/technote-ocsp-cert-revocation.md:57]:
```suggestion
The next two mechanisms were mainly introduced for browsers and are not directly supported by
```

File: Documentation/technote-ocsp-cert-revocation.md [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+the Java JDK.
+
+**OCSP Must-Staple**. OCSP has one failing in that many clients will Soft-Fail, i.e. allow connections
+to proceed if they do not receive an OCSP response in time. This leads toa false sense of security as
```
Comment (cocreature) [inline Documentation/technote-ocsp-cert-revocation.md:61]:
```suggestion
to proceed if they do not receive an OCSP response in time. This leads to a false sense of security as
```

File: Documentation/technote-ocsp-cert-revocation.md [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+clients will not fail closed. OCSP Must-Staple is a new Certificate extension that the
```
Comment (cocreature) [inline Documentation/technote-ocsp-cert-revocation.md:62]: closed?

File: Documentation/technote-ocsp-cert-revocation.md [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+
+**Chrome CRLSet / Firefox OneCRL**. These are browser specific solutions to the
+revocation issue. The browser vendors collate a list of CRLs for major Certificate Authority certificates (the CA
+not the end leaf certificates) and distribute as part of their browser package. This resolves the download issue
+but does not support other CAs or end-leaf certificates. It is focused on revoking CAs that
+are found to be compromised.
+
+**Other mechanisms**
+
+Another mechanism tat can be used is to issue certificates with short lifetimes. Connections are
```
Comment (cocreature) [inline Documentation/technote-ocsp-cert-revocation.md:75]:
```suggestion
Another mechanism that can be used is to issue certificates with short lifetimes. Connections are
```

File: Documentation/technote-ocsp-cert-revocation.md [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+For the JDK, you can use the jSSLKeyLog agent to capture the TLS session key to a file and then configure Wireshark to
+read. Wireshark is then able to decrypt the traffic, which is required to see some packate details (for example, the
+OCSP Staple response details)
+```aidl
+-javaagent:$ROOTDIR/jSSLKeyLog.jar=jssl-key.log 
+```
+
+Similarly, it is also possible to trace the GRPC/Protobuf traffic by configuring the
+Protobuf add-in Wireshark to use the Daml protobuf files. You may need to download Daml
+repo and Google base grpc protobuf files for this to work successfully.
```
Comment (cocreature) [inline Documentation/technote-ocsp-cert-revocation.md:139]: Worth mentioning that this is probably a bad idea for performance, security, … and should only be used for debugging?

File: Documentation/technote-ocsp-cert-revocation.md [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+./run-tls-server.sh
+
+# Run a TLS Client
+# Note you can switch between standard TLS and Mutual TLS clients
+# Client should show contents of the test.txt file in the server directory 
+./run-tls-client.sh
+```
+
+What do the Java parameters do?
+```aidl
```
Comment (cocreature) [inline Documentation/technote-ocsp-cert-revocation.md:190]: more aidl

File: Documentation/technote-ocsp-cert-revocation.md [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+- revoking the cert
+- access ledger again and see access failure from revoked cert
+
+## Daml Server-Side OCSP Stapling
+
+The default TLS Provider used by Daml is BoringSSL. This is a simplified implementation that
+reduces the TLS code base and removes a variety of features, particularly those that are not
+used frequently. This includes OCSP Stapling.
+
+It is possible to reconfigure Daml to use the installed JDK in preference to
```
Comment (cocreature) [inline Documentation/technote-ocsp-cert-revocation.md:254]: "requires a custom build" means it’s not possible for users.

**Review by cocreature — COMMENTED** [review]
Mostly lgtm with a few comments
